src/toil/jobStores/aws/jobStore.py

```diff
@@ -1153,39 +1185,122 @@ def __getBucketVersioning(self, bucket):
-            for attempt in retry_s3():
-                with attempt:
-                    for upload in self.filesBucket.list_multipart_uploads():
-                        upload.cancel_upload()
-            if self.__getBucketVersioning(self.filesBucket) in (True, None):
+        self.cleanJobStore('aws:%s:%s' % (self.region, self.namePrefix))
+
+    @classmethod
+    def cleanJobStore(cls, locator):
+        info = cls._processLocator(locator)
+        filesBucket = None
+
+        # First we load all job store components.
+        try:
+            with closing(cls._connectS3(info.region)) as s3:
                 for attempt in retry_s3():
                     with attempt:
-                        for key in list(self.filesBucket.list_versions()):
-                            self.filesBucket.delete_key(key.name, version_id=key.version_id)
+                        filesBucket = s3.get_bucket('%s%sfiles' %
+                                                    (info.namePrefix, cls.nameSeparator))
+        except S3ResponseError as e:
+            if e.error_code == 'NoSuchBucket':
+                pass
             else:
-                for attempt in retry_s3():
-                    with attempt:
-                        for key in list(self.filesBucket.list()):
-                            key.delete()
-            for attempt in retry_s3():
-                with attempt:
-                    self.filesBucket.delete()
-        for domain in (self.filesDomain, self.jobsDomain):
-            if domain is not None:
-                for attempt in retry_sdb():
-                    with attempt:
-                        domain.delete()
+                raise
+
+        def getDomain(name):
+            # no such domain predicate has been removed to prevent unnecessary retries
+            predicate = (lambda e: sdb_unavailable(e) or
+                                   connection_reset(e) or
+                                   retryable_ssl_error(e))
+            try:
+                with closing(cls._connectSimpleDB(info.region)) as sdb:
+                    for attempt in retry_sdb(predicate=predicate):
+                        with attempt:
+                            return sdb.get_domain(name)
+            except SDBResponseError as e:
+                if e.error_code == 'NoSuchDomain':
+                    return None
+                else:
+                    raise
+
+        filesDomain = getDomain('%s%sfiles' % (info.namePrefix, cls.nameSeparator))
+        jobsDomain = getDomain('%s%sjobs' % (info.namePrefix, cls.nameSeparator))
+        registry = getDomain('toil-registry')
+        if registry is None:
+            log.warn('Toil registry domain does not exist. Nothing will be updated after clean.')
 
+        # Now all variables pointing to job store components are None if and only if they do not
+        # exist.
+        i = 0
+        while True:
+            if filesBucket == filesDomain == jobsDomain is None:
```

Comment:
Make this dynamic. The central pattern is that you create a set of domains to delete. Once you successfully delete a domain, remove it from the set. Repeat until the set is empty.